File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
#
# @lc app=leetcode id=2058 lang=python3
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

from typing import Optional, List

class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        critical = []
        if not head or not head.next:
            return [-1, -1]
        prev = head
        cur = head.next
        pos = 2
        while cur.next:
            nxt = cur.next
            if (cur.val > prev.val and cur.val > nxt.val) or \
               (cur.val < prev.val and cur.val < nxt.val):
                critical.append(pos)
            prev = cur
            cur = nxt
            pos += 1
        if len(critical) < 2:
            return [-1, -1]
        min_d = float('inf')
        for j in range(1, len(critical)):
            d = critical[j] - critical[j - 1]
            min_d = min(min_d, d)
        max_d = critical[-1] - critical[0]
        return [min_d, max_d]

# @lc code=end

File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
#
# @lc app=leetcode id=2058 lang=cpp
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#
# @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        if (!head || !head->next || !head->next->next) {
            return {-1, -1};
        }
        
        int firstCritical = -1;
        int prevCritical = -1;
        int minDistance = INT_MAX;
        
        ListNode* prev = head;
        ListNode* curr = head->next;
        int position = 1;
        
        while (curr->next) {
            ListNode* next = curr->next;
            
            // Check if current node is a critical point
            if ((curr->val > prev->val && curr->val > next->val) ||
                (curr->val < prev->val && curr->val < next->val)) {
                
                if (firstCritical == -1) {
                    firstCritical = position;
                } else {
                    minDistance = min(minDistance, position - prevCritical);
                }
                prevCritical = position;
            }
            
            prev = curr;
            curr = next;
            position++;
        }
        
        // If fewer than 2 critical points
        if (firstCritical == -1 || prevCritical == firstCritical) {
            return {-1, -1};
        }
        
        int maxDistance = prevCritical - firstCritical;
        
        return {minDistance, maxDistance};
    }
};
# @lc code=end